typeck: use NoExpectation to check return type of diverging fn

durka opened the pull request:
Fixes #35849.

Thanks @eddyb.

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

eddyb commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -708,7 +708,12 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
+                                      NoExpectation
+                                  } else {
+                                      ExpectHasType(fcx.ret_ty)
+                                  });
```
I'd put this in a variable or call a different version of `check_block` in the `TyNever` case.

eddyb commented on src/test/run-pass/diverging-fn-tail-35849.rs:
```diff
@@ -0,0 +1,18 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
```
Nice test file naming scheme! cc @rust-lang/compiler Can we decide on something like this when the testcase comes from an issue but it has some kind of categorization otherwise?

eddyb commented:
r=me with nit fixed.

durka commented:
Nit fixed. This also needs to be `beta-nominated`.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 9d53faf has been approved by `eddyb`

<!-- @bors r=eddyb 9d53faf1ea8d207ab737e8e0a7c3fafb14ed0754 -->

canndrew commented:
This won't allow something like `fn foo() -> ! { 3 }` to compile will it?

eddyb commented:
@canndrew It's the same logic as before, I believe liveness is what prevents control-flow from exiting the function if the function is supposed to be divergent (in a more sound manner).

durka commented:
@canndrew @eddyb yep we got a problem here:

    [root@li1424-173 gh35849]# build/x86_64-unknown-linux-gnu/stage1/bin/rustc - -o bad <<<'fn foo() -> ! { 42 } fn main() { foo(); }'
    warning: broken MIR (return = const 42i32): bad assignment (! = i32): Sorts(ExpectedFound { expected: !, found: i32 })
     --> <anon>:1:17
      |
    1 | fn foo() -> ! { 42 } fn main() { foo(); }
      |                 ^^
    
    [root@li1424-173 gh35849]# ./bad
    Segmentation fault (core dumped)

eddyb commented:
Whoa how did that happen?! That's bad, because the code erroring before wasn't in typeck AFAIK.

@bors r-

durka commented:
IRC investigation reveals that #35162 removed an important check from the liveness pass, which I will reinstate whenever LLVM finishes building. Now I'm wondering if that PR removed anything else important.

canndrew commented:
Are you talking about the check in `check_ret` which was removed? If so, I think I was assuming that any block which diverges would get typed `!` (including ones ending in an expression). Would it better to fix that than to make a special case for `!` not having to be type-check correctly and then make another special case for it further down the pipeline?

eddyb commented:
@canndrew Maybe that can work, but we'll *always* need a sanity check - if typeck ends up being able to handle divergence correctly, we'd just have a *proper* CFG-based check ICE instead of erroring.

arielb1 commented:
Maybe just add a "redundant" return type check always?

canndrew commented:
I'm generally uncomfortable making special cases for `!` - part of the motivation for this RFC is that a lot of this sort of thing can be handled naturally by the type system. Plus `!` isn't the only diverging type.

eddyb commented:
@canndrew The special cases need to be there until there's a RFC that says `[x: !]` is typed `!` instead of `[!; 1]` *or* that `[!; 1]` is-a-subtype-of/coerces-to `!`, and until the definition of "diverging function" changes.
We risk too much if we don't have sanity checks for the CFG of a diverging function - the MIR warning being one of them, we should probably make those warnings ICE now.

durka commented:
> we should probably make those warnings ICE now

cf #35499

canndrew commented:
I don't think that liveness should have special cases for `!` at all. As far as its is concerned `!` isn't a special case, it's just one of many uninhabited types. If liveness can handle `Void` it should be able to handle `!` as well. The only reason it can't is that we just added another special case for `!` which broke typechecking and allowed clear nonsense like `fn foo() -> ! { 3 }` to get through. It's one special case begetting another.

> The special cases need to be there until there's a RFC that says `[x: !]` is typed ! instead of `[!; 1]` or that `[!; 1]` is-a-subtype-of/coerces-to `!`

Why? This can get typechecked correctly:

```rust
fn foo() -> ! {
    [panic!()];
}
```

So there's some logic that can see the `!` in the middle of that expression. Can we make that logic run on the block's final expression as well and not just it's statements?

eddyb commented:
@canndrew No, that doesn't get type-checked *either*. The error you get when the function is declared to return `T` and you have a semicolon and no divergence is `not all control paths return a value`, *also* from liveness.
That is, if you don't have a trailing expression, it's *not* checked at all, although this seems to only happen at the outermost level, a type error is produced if you nest blocks.

durka commented:
Amended to restore the special case in `Liveness::check_ret`.

eddyb commented on src/librustc/middle/liveness.rs:
```diff
@@ -1479,43 +1479,50 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
+                        "computation may converge in a function marked as diverging");
+            }
+        } else {
+            if self.live_on_entry(entry_ln, self.s.no_ret_var).is_some() {
```
This could be an `else-if` to make the diff simpler.

eddyb commented on src/test/compile-fail/diverging-fn-tail-35849.rs:
```diff
@@ -0,0 +1,22 @@
+    unsafe {
+        ::std::mem::transmute::<f64, [u8; 8]>(panic!())
+    }
+}
```
I think this should be in a different test, to make sure compiler pass ordering doesn't hide bugs.

durka commented:
Nits knitted.

eddyb commented on src/librustc/middle/liveness.rs:
```diff
@@ -1479,7 +1479,10 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
                 &self.fn_ret(id));
 
-        if self.live_on_entry(entry_ln, self.s.no_ret_var).is_some() {
+        if fn_ret.is_never() && self.live_on_entry(entry_ln, self.s.clean_exit_var).is_some() {
```
These should not be combined, i.e. the `else` below should be to `fn_ret.is_never()` alone, to match the old semantics.

durka commented:
Knit nitted.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit ddebea0 has been approved by `eddyb`

<!-- @bors r=eddyb ddebea05d5b4f1f76e985fc55a655af29b2f0edf -->

durka commented:
Fixed the travis failure and added some FIXME comments.

We unfortunately needed to ignore a test added for `#![feature(never_type)]` because it can't be done with the current structure of the liveness code. A fix may be possible, but the consensus was that we shouldn't try to get it into this PR, to keep it backportable. I'll file separate bugs.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 702ea71 has been approved by `eddyb`

<!-- @bors r=eddyb 702ea7169cc20f3e572576e68339edc0b8fc5939 -->

bors commented:
:hourglass: Testing commit 702ea71 with merge 5f31fda...

alexcrichton commented:
@bors: retry force clean

* restarted buildbot

bors commented:
:hourglass: Testing commit 702ea71 with merge 03e23c7...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10275), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10329), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1564), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1389), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3548), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10269), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10231), [auto-linux-64-opt-no-mir](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-no-mir/builds/198), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2203), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10208), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3432), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5380), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10307), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10280), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2226), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1404), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5395), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2256), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5371), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4562), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1583), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5376), [auto-win-msvc-64-opt-no-mir](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-no-mir/builds/207), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2251)
Approved by: eddyb
Pushing 03e23c7f9a10b18e1687bfd92c6807e9342aa00b to master...
